File: ios/Classes/readers/configurations/SubjectConfigurationReader.swift
// Copyright (c) 2022-present Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0,
// and you may not use this file except in compliance with the Apache License Version 2.0.
// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the Apache License Version 2.0 is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

import Foundation
import SnowplowTracker
import CoreImage

struct SubjectConfigurationReader: Decodable {
    let userId: String?
    let networkUserId: String?
    let domainUserId: String?
    let userAgent: String?
    let ipAddress: String?
    let timezone: String?
    let language: String?
    let screenResolution: [Double]?
    let screenViewport: [Double]?
    let colorDepth: Double?

    var screenResolutionSize: SPSize? {
        if let screenResolution = self.screenResolution,
           let width = screenResolution.first,
           let height = screenResolution.last {
            return SPSize(width: Int(width), height: Int(height))
        }
        return nil
    }
    var screenViewportSize: SPSize? {
        if let screenViewport = self.screenViewport,
           let width = screenViewport.first,
           let height = screenViewport.last {
            return SPSize(width: Int(width), height: Int(height))
        }
        return nil
    }
}

extension SubjectConfigurationReader {
    func toConfiguration() -> SubjectConfiguration {
        let configuration = SubjectConfiguration()
        
        if let uid = self.userId { configuration.userId(uid) }
        if let nid = self.networkUserId { configuration.networkUserId(nid) }
        if let did = self.domainUserId { configuration.domainUserId(did) }
        if let ua = self.userAgent { configuration.useragent(ua) }
        if let ip = self.ipAddress { configuration.ipAddress(ip) }
        if let tz = self.timezone { configuration.timezone(tz) }
        if let lang = self.language { configuration.language(lang) }
        if let sr = self.screenResolutionSize { configuration.screenResolution(sr) }
        if let sv = self.screenViewportSize { configuration.screenViewPort(sv) }
        if let cd = self.colorDepth { configuration.colorDepth(NSNumber(value: Int(cd))) }
        
        return configuration
    }
}

File: android/src/main/kotlin/com/snowplowanalytics/snowplow_tracker/readers/configurations/SubjectConfigurationReader.kt
// Copyright (c) 2022 Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0,
// and you may not use this file except in compliance with the Apache License Version 2.0.
// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the Apache License Version 2.0 is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

package com.snowplowanalytics.snowplow_tracker.readers.configurations

import com.snowplowanalytics.snowplow.configuration.SubjectConfiguration
import com.snowplowanalytics.snowplow.util.Size

class SubjectConfigurationReader(val values: Map<String, Any>) {
    private val valuesDefault = values.withDefault { null }

    val userId: String? by valuesDefault
    val networkUserId: String? by valuesDefault
    val domainUserId: String? by valuesDefault
    val userAgent: String? by valuesDefault
    val ipAddress: String? by valuesDefault
    val timezone: String? by valuesDefault
    val language: String? by valuesDefault
    val screenResolution: Pair<Double, Double>? by valuesDefault
    val screenViewport: Pair<Double, Double>? by valuesDefault
    val colorDepth: Double? by valuesDefault

    val screenResolutionSize: Size? by lazy {
        screenResolution?.let {
            val screenWidth = it.first.toInt()
            val screenHeight = it.second.toInt()
            Size(screenWidth, screenHeight)
        }
    }
    val screenViewportSize: Size? by lazy {
        screenViewport?.let {
            val screenVPWidth = it.first.toInt()
            val screenVPHeight = it.second.toInt()
            Size(screenVPWidth, screenVPHeight)
        }
    }

    fun toConfiguration(): SubjectConfiguration {
        val subjectConfig = SubjectConfiguration()

        userId?.let { subjectConfig.userId(it) }
        networkUserId?.let { subjectConfig.networkUserId(it) }
        domainUserId?.let { subjectConfig.domainUserId(it) }
        userAgent?.let { subjectConfig.useragent(it) }
        ipAddress?.let { subjectConfig.ipAddress(it) }
        timezone?.let { subjectConfig.timezone(it) }
        language?.let { subjectConfig.language(it) }
        screenResolutionSize?.let { subjectConfig.screenResolution(it) }
        screenViewportSize?.let { subjectConfig.screenViewPort(it) }
        colorDepth?.let { subjectConfig.colorDepth(it.toInt()) }

        return subjectConfig
    }
}
